File: Python/majority_element.py
# Uses python3
import sys

def get_majority_element(arr):


    if len(arr) < 2:
        return arr[0]

    midpoint = len(arr) // 2


    left = arr[:midpoint]


    right = arr[midpoint:]


    majority_left = 0
    majority_right = 0


    temp_left = get_majority_element(left)
    if temp_left != None:
        majority_left = temp_left



    temp_right = get_majority_element(right)
    if temp_right != None:
        majority_right = temp_right


    counterLeft = 0
    counterRight = 0

    for x in range( len(arr) ):


        if arr[x] == majority_left:
            counterLeft+=1
        elif arr[x] == majority_right:
            counterRight+=1

    if counterLeft > len(arr) // 2:
        print("hello")
        return 1

    elif counterRight > len(arr) // 2:
        print("yellow")
        return 1

    else:
        return 0


input = sys.stdin.read()
n, *arr = list(map(int, input.split()))


left = []
right = []
temp = []


result = get_majority_element(arr)
if result > 0:
    result = 1

print(result)

# if get_majority_element(arr,left, right) != -1:
#     print(1)
# else:
#     print(0)



# I.e. say you have array a = [2 3 9 2 2] as stated in the problem statement. You split it into b1 = [2 3] and b2 = [9 2 2]. You then split b1 into b11 = [2] and b12 = [3]. You then return 2 and 3, and count their occurrence in b1. If either happens more than (n/2 = 2/2 = 1) times in b1, they are a majority element of b1. Otherwise, b1 contains no majority element and you return 0. Your program will then start recursively splitting b2 using the same procedure as above. It will then see that a majority element in b2 is 2, and count the number of its occurrences in a, to conclude that it happens more than (n/2 = 5/2 = 2) times in a and it is thus a majority element.

File: JavaScript/Divide and Conquer/majorityElement.js

var n = 0
var input = []
var lineNum = 1

var readline = require('readline');
var rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,

});

rl.on('line',function(cmd){
  if(lineNum === 1){
    n = parseInt(cmd)
  } else{
    input = cmd.split(' ').map(Number)
  }
  lineNum += 1


})


rl.on('close',function(cmd){
  var result = get_majority_element(input,0,input.length-1)
  if(result > 0){
    console.log(1)
  } else{
    console.log(0)
  }

  rl.close()
})

function get_majority_element(arr, left, right){

    if (left === right)
        return -1
    if (left + 1 === right)
        return arr[left]

    var mid = left + Math.floor( (right - left) / 2)

    var majorityLeft = get_majority_element(arr,left,mid)

    var majorityRight = get_majority_element(arr,mid,right)

    var resultMajority = isMajorityElement(arr.slice(left,right + 1), majorityLeft,majorityRight)
    if(resultMajority > 0){
      return resultMajority
    }


    return -1


}

function isMajorityElement(arr,left,right){
  var count = 0
  var countLeft = 0
  var countRight = 0
  for(var x = 0; x < arr.length; x++){
    if(arr[x] === left){
      countLeft += 1
    }
    else if(arr[x] === right){
      countRight += 1
    }
    if(countLeft > ( arr.length / 2) )
      return left
    if (countRight > (arr.length / 2))
      return right
  }
    return -1
}
